Ullaakut/Bloggo
# Login & Registration API (#54) (#65)

## Goal of this PR

This PR aims at replacing the original authorization system that relied on tokens that were created using OpenID Connect with a registration & login API that will allow for users to use Bloggo to create and access their accounts without the need for any third party.

## Implementation

* Adds a `User` model
* Registering is done using the new `/register` route. User info is sent through a JSON object. For now the password is sent in clear and isn't hashed, and the API is available only through HTTP not HTTPS, so this is not yet secure.
* Logging in is done using the new `/login` route. User email and password is sent through a JSON object, and compared to the email/password combination stored in the database. Again, this isn't secure yet as the password has to be sent in clear through HTTP.
* Modifies the UserRepository to make it more generic by allowing to filter `Users` using any of their fields and not just the `ID`. This was necessary for getting users from their email addresses.
* Updates Blueprint documentation
* Non-admin accounts will be used later on to leave comments and likes

## Todo

- [x] Update functional tests
  - [x] Login API
  - [x] Registration API
  - [x] Authorization middleware
  - [x] Blog Post API
- [x] Update unit tests to get back to 100% coverage
  - [x] Test token service
  - [x] Test user controller
- [x] Update postman collection
  - [x] Add login route
  - [x] Add register route
- [x] Update migration files
  - [x] Remove default users
- [x] Update documentation to describe this and explain how to create the users
- [x] Make it so that only one user can be admin at a time
- [x] The frontend will have to set the `IsAdmin` field to true when creating the admin user in the setup phase of the blog
- [x] Prevent from having multiple users with the same email address

## How to test it

* Using the postman collection
* `docker-compose up`
* Try creating a non admin account, should work
* Try logging in the account, should work
* Use the token that you get for reading blog posts, should work
* Use the token that you get for creating blog posts, should not work as it isn't admin
* Try creating an admin account, should work
* Try logging in the admin account, should work
* Use the admin token for reading blog posts, should work
* Use the admin token for creating blog posts, should work
* Try creating another admin account, should not work

## Screenshots

<img width="644" alt="screenshot 2018-08-26 at 21 57 26" src="https://user-images.githubusercontent.com/6976628/44632456-0a14b180-a97b-11e8-9bc1-33feb925c954.png">
